feat: Blocking provider mutator

znonthedev opened the pull request:
## This PR
Adds setProviderAndWait method in the API that will set the provider synchronously  

- adds this new feature

### Related Issues

Fixes #240

### Follow-up Tasks
add test scenarios
update readme

znonthedev commented:
Hi @toddbaert, could you please review this PR? I am very new to Go and not sure if this implementation is correct. I will refactor it and keep it DRY if the current implementation meets the expectations outlined in the issue. Thanks in advance.

toddbaert commented:
Thank you @znonthedev !

I will review, and also add some reviewers who are more experienced with Go than me! Many people are on holiday however, so it may take a few days to get some approvals. Your patience is appreciated! :pray:

Kavindu-Dodan commented on openfeature/api.go:
```diff
@@ -211,6 +235,46 @@ func (api *evaluationAPI) initNewAndShutdownOld(newProvider FeatureProvider, old
+func (api *evaluationAPI) initNewAndShutdownOldSync(newProvider FeatureProvider, oldProvider FeatureProvider) {
+	v, ok := newProvider.(StateHandler)
+	if ok && v.Status() == NotReadyState {
+		err := v.Init(api.evalCtx)
+		if err != nil {
+			api.eventExecutor.eventChan <- eventPayload{
+				Event{
+					ProviderName:         newProvider.Metadata().Name,
+					EventType:            ProviderError,
+					ProviderEventDetails: ProviderEventDetails{},
+				}, newProvider,
+			}
+		} else {
+			api.eventExecutor.eventChan <- eventPayload{
+				Event{
+					ProviderName:         newProvider.Metadata().Name,
+					EventType:            ProviderReady,
+					ProviderEventDetails: ProviderEventDetails{},
+				}, newProvider,
+			}
+		}
```
This is duplicate of the same calls of https://github.com/open-feature/go-sdk/blob/main/openfeature/api.go#L174-L194 

Instead of duplication, you can extract this code to a dedicated function (ex:- `func initHelper(...)`) . Then you can call this function as a go routine for async operation (L174-L194 ) and as a direct function call for sync operation.

znonthedev replied:
Refactored and avoids duplication

Kavindu-Dodan commented on openfeature/openfeature.go:
```diff
@@ -23,6 +23,10 @@ func SetProvider(provider FeatureProvider) error {
 	return api.setProvider(provider)
 }
 
+func SetProviderAndWait(provider FeatureProvider) error {
```
+1 

I am fine with this method of naming. Please add a doc comment explaining the functionality. Also, please update `SetProvider` (L20-L24) doc comment mention that it is asynchronous (non-blocking)

znonthedev replied:
Added doc comments

znonthedev commented:
I have updated the PR based on the comments. I have added two test cases also. I am not sure on the test cases. Do I need to add more test cases or change the current ones

toddbaert commented:
We need to add an equivalent function for setting a named provider and waiting (see [here](https://github.com/open-feature/go-sdk/pull/244/files#diff-434e8d62e6a56910999195bb04f7c78e5373efd76288dc4039cd58686c99ae8bR94)).

toddbaert commented:
@bacherfl when you get a moment, could you take a look at this as well?

znonthedev commented:
> We need to add an equivalent function for setting a named provider and waiting (see [here](https://github.com/open-feature/go-sdk/pull/244/files#diff-434e8d62e6a56910999195bb04f7c78e5373efd76288dc4039cd58686c99ae8bR94)).

Certainly, @toddbaert. I had the same thought. I will incorporate the necessary changes into the PR

toddbaert commented on openfeature/client_test.go:
```diff
@@ -749,6 +749,87 @@ func TestRequirement_1_4_13(t *testing.T) {
 
 // TODO Requirement_1_6_1
+
+func TestWaitForProvider(t *testing.T) {
```
It might be nice to add a test that works like this:

- set a provider which is ready after 1s (has an init function that returns after 1s)
- returns a valid flag value after that time only
- returns an error before 1s

This is consistent with how we'd expect many providers to work. Maybe you could add some options to `NewMockFeatureProvider` that would support this?

znonthedev replied:
okay, I got the idea. I'll give it a try, but since I'm new to Go, it might take me some time 😇 .

bacherfl replied:
I just had a look at the test, and to make sure the `Init` function actually gets called, you will need to have a type that consolidates a `MockFeatureProvider` and `MockStateHandler` - otherwise the check [here](https://github.com/open-feature/go-sdk/pull/244/files#diff-434e8d62e6a56910999195bb04f7c78e5373efd76288dc4039cd58686c99ae8bR270) will fail and the function will proceed without calling the `Init` function of the provider.

Since gomock, which we use for creating the mocks seems to not support the generation of a type that implements multiple interfaces, i believe the simplest way to get a struct that implements both the FeatureProvider and the StateHandler interfaces is to consolidate them into a single struct, such as follows:

```go
type mockProviderAndStateHandler struct {
	*MockFeatureProvider
	*MockStateHandler
}
```

This type should be defined in the `client_test.go` file, and can then be used to pass it to the `SetProviderAndWait` function, such as in the following example:

```
...
		mockProvider.EXPECT().Metadata().AnyTimes()
		mockProvider.EXPECT().Hooks().AnyTimes()
		mockProvider.EXPECT().IntEvaluation(context.Background(), flagKey, defaultValue, flatCtx).
			Return(IntResolutionDetail{
				Value: 0,
				ProviderResolutionDetail: ProviderResolutionDetail{
					ResolutionError: NewGeneralResolutionError("test"),
				},
			}).Times(2)

		mockStateHandler := NewMockStateHandler(ctrl)

                // TODO add the desired behavior for the Init function of the MockStateHandler

		o := & mockProviderAndStateHandler{
			MockFeatureProvider: mockProvider,
			MockStateHandler:    mockStateHandler,
		}

		err := SetProviderAndWait(o)
...
```

Kavindu-Dodan commented on openfeature/__debug_bin1610825419:
please remove this file

Kavindu-Dodan commented on openfeature/api.go:
```diff
@@ -166,32 +216,35 @@ func (api *evaluationAPI) forTransaction(clientName string) (FeatureProvider, []
+// initNewAndShutdownOldAsync is a helper to initialise new FeatureProvider and shutdown the old FeatureProvider.
 // Operations happen concurrently.
-func (api *evaluationAPI) initNewAndShutdownOld(newProvider FeatureProvider, oldProvider FeatureProvider) {
+func (api *evaluationAPI) initNewAndShutdownOldAsync(newProvider FeatureProvider, oldProvider FeatureProvider) {
```
You may refactor to combine `initNewAndShutdownOldAsync` & `initNewAndShutdownOldSync` as they perform similar set of operations with only difference being sync vs async.

bacherfl commented on openfeature/api.go:
```diff
@@ -49,7 +49,31 @@ func (api *evaluationAPI) setProvider(provider FeatureProvider) error {
+
+// setProviderAndWait sets the default FeatureProvider of the evaluationAPI. Returns an error if FeatureProvider is nil
+// This is a blocking call and will wait for the provider to be ready
+func (api *evaluationAPI) setProviderAndWait(provider FeatureProvider) error {
```
Nitpick: to stay consistent with the naming i would suggest to call this function `setProviderSync`

toddbaert replied:
Agreed. The spec doesn't dictate function names. I think your suggestion makes sense here.

Kavindu-Dodan commented:
@znonthedev are you planning to complete this contribution soon ? There are pending changes in some providers which require this feature.

toddbaert commented:
Closed with https://github.com/open-feature/go-sdk/pull/251
